### PR TITLE
Correct and clarify the handling of empty/zero-length `Doc`s during training and inference

### PR DESCRIPTION
**Pull request by shadeMe**
## Description

This PR addresses two issues:
- When zero-length inputs passed to the transformer pipe, the outputs yielded by an attached listener has the wrong shape in the representation width dimension. This is more of a question of correctness than a practical concern due to the lack of tokens in such `Doc`s, but we explicitly clarify why we use zero-width outputs.
- Backprop callbacks and gradients were not correctly aligned to their corresponding model outputs (outputs from the transformer pipe) when empty documents occurred in the middle of a training batch. The existing tests for this corner-case were inadvertently passing due to the behaviour of `zip`, i.e., the test always added the empty document to the end of the batch, and the subsequent `zip` operation would implicitly skip the last backprop/gradient due to the latter list being shorter than the list of model outputs.

It also adds comments about the pre-conditions to clarify the circumstances under which zero-length inputs are passed to the model.

### Types of change

Bug Fix

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: spacy_transformers/layers/trfs2arrays.py [hunk 1]
```diff
@@ -21,33 +22,65 @@ def forward(model: Model, trf_datas: List[TransformerData], is_train: bool):
     grad_factor = model.attrs["grad_factor"]
     outputs = []
     backprops = []
+
+    # If possible, cache the width for allocating zero-length tensors.
```
Comment (danieldk) [inline spacy_transformers/layers/trfs2arrays.py:26]: Hmmm, I am not a big fan of this, since it is only conditionally right. I think it's nicer to be predictable and be right or wrong all the time.

Luckily, this won't be an issue in curated transformers.

File: spacy_transformers/layers/trfs2arrays.py [hunk 2]
```diff
@@ -21,33 +22,65 @@ def forward(model: Model, trf_datas: List[TransformerData], is_train: bool):
...
             d_model_output = ModelOutput(
                 last_hidden_state=model.ops.alloc(
                     trf_data.model_output.last_hidden_state.shape,  # type: ignore
                     dtype=trf_data.model_output.last_hidden_state.dtype,  # type: ignore
                 )
             )
-            d_dst = get_d_dst(d_output)
-            d_src = get_d_src(d_dst)
+            d_dst = get_d_dst(d_output)  # type: ignore
+            d_src = get_d_src(d_dst)  # type: ignore
```
Comment (danieldk) [inline spacy_transformers/layers/trfs2arrays.py:83]: Do we need the ignores if we type-assert (or cast) after the condition that skips the loop iteration?

File: spacy_transformers/layers/trfs2arrays.py (hunk 1 above)
Comment (shadeMe) [reply spacy_transformers/layers/trfs2arrays.py:26]: Hmm, fair point. I'll leave the width at zero in all cases then.

**Review by danieldk — APPROVED** [review]

**Comment by danieldk** [comment]
Assigned to you Sofie, since you added the handling of zero-length docs, so it would be good get an additional sanity check.

File: spacy_transformers/layers/trfs2arrays.py [hunk 3]
```diff
@@ -21,33 +22,64 @@ def forward(model: Model, trf_datas: List[TransformerData], is_train: bool):
...
+                # This is the general case for non-zero length documents.
                 src = model.ops.reshape2f(tensor_t_i, -1, trf_data.width)  # type: ignore
                 dst, get_d_src = apply_alignment(model.ops, trf_data.align, src)
                 output, get_d_dst = pooling(dst, is_train)
                 outputs.append(output)
-                backprops.append((get_d_dst, get_d_src))
+                backprops.append((get_d_dst, get_d_src))  # type: ignore
         else:
-            outputs.append(model.ops.alloc2f(0, 0))
+            # This can happen during prediciton/training for zero-length documents. Since zero-length docs
```
Comment (svlandeg) [inline spacy_transformers/layers/trfs2arrays.py:51]:
```suggestion
            # This can happen during prediction/training for zero-length documents. Since zero-length docs
```

**Review by svlandeg — APPROVED** [review]
Nice digging, looks great!
